proxy/coco/src/request.rs

```diff
@@ -0,0 +1,336 @@
+// is changing during state transitions.
+// See https://github.com/rust-lang/rust-clippy/issues/4859 for more information.
+#![allow(clippy::use_self)]
+
+use std::{
+    collections::HashMap,
+    ops::{Deref, Sub},
+};
+
+use either::Either;
+use serde::{Deserialize, Serialize};
+
+use librad::{net::peer::types::Gossip, peer::PeerId, uri::RadUrn};
+
+pub mod states;
+pub use states::*;
+pub mod existential;
+pub use existential::SomeRequest;
+pub mod waiting_room;
+
+/// Private trait for sealing the traits we use here.
+mod sealed;
+
+/// A `Request` represents the lifetime of requesting an identity in the network via its
+/// [`RadUrn`].
+///
+/// The `Request`'s state is represented by the `S` type parameter. This parameter makes sure that
+/// a `Request` transitions through specific states in a type safe manner.
+///
+/// These transitions are pictured below:
+///
+/// ```text
+///      +----------------------------------v
+///      |                             +---------+
+///      |                   +-------->+cancelled+<------+
+///      |                   |         +----+----+       |
+///      |                   |              ^            |
+///      |                   |              |            |
+/// +----+----+       +------+--+       +---+-+      +---+---+       +------+
+/// | created +------>+requested+------>+found+----->+cloning+------>+cloned|
+/// +---------+       +------+--+       +--+--+      +---+---+       +------+
+///                          |  ^-------+  |  ^------+   |
+///                          |    failed   |   failed    |
+///                          |             v             |
+///                          |          +--+------+      |
+///                          +--------->+timed out+------+
+///                                     +---------+
+/// ```
+///
+/// The `T` type parameter represents some timestamp that is chosen by the user of the `Request`
+/// API. Note that it makes it easy to test by just choosing `()` for the timestamp.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Request<S, T> {
+    /// The identifier of the identity on the network.
+    urn: RadUrn,
+    /// The number of attempts this request has made to complete its job.
+    attempts: Attempts,
+    /// The timestamp of the latest action to be taken on this request.
+    timestamp: T,
+    /// The state of the request, as mentioned above.
+    state: S,
+}
+
+impl<S, T> Deref for Request<S, T> {
+    type Target = S;
+
+    fn deref(&self) -> &Self::Target {
+        &self.state
+    }
+}
+
+impl<S, T> From<Request<S, T>> for Gossip {
+    fn from(request: Request<S, T>) -> Self {
+        Self {
+            urn: request.urn,
+            rev: None,
+            origin: None,
+        }
+    }
+}
+
+impl<S, T> Request<S, T> {
+    /// Get the [`RadUrn`] that this `Request` is searching for.
+    pub const fn urn(&self) -> &RadUrn {
+        &self.urn
+    }
```

Comment:
TIL on `pub const`

Out of curiosity, why having this getter instead of making `urn` public?

Reply:
I think it'd be disastrous if a user could change the `RadUrn` of the request in the `WaitingRoom`. #eNcApSuLaTiOn

Reply:
would that require it to be `mut`?

Reply:
Ya, it would I believe.